### PR TITLE
manifest2proto: Added type to handle connection instance (contained in particle instance within a recipe).

### PR DESCRIPTION
**Pull request by alxrsngrtn**
This is the first in a series of PRs to add types to handle connection instances. Here, we update the proto definition and (TS) serialization. The next PR will implement the proto deserialization into Kotlin.

### CODE REVIEW

File: src/tools/tests/manifest2proto-test.ts [hunk 1]
```diff
@@ -182,6 +193,187 @@ describe('manifest2proto', () => {
...
+  it('encodes variable particle instance types', async () => {
+    const manifest = await Manifest.parse(`
+      particle Writer1
+        first: writes [Foo {a: Text, c: Text}]
+        second: writes [Bar {a: Text, b: Text}]
+      particle Writer2
+        first: writes [Foo {a: Text, b: Text}]
+        second: writes [Bar {a: Text, c: Text}]
+      particle Reader1
+        first: reads [~f with Foo {a: Text, b: Text, c: Text}]
```
Comment (piotrswigon) [inline src/tools/tests/manifest2proto-test.ts:255]: Something is fishy here. The type of a handle is at most `[Foo {a: Text}]`, this particle should not typecheck.

Interestingly though, the encoded type is correct.

File: src/tools/tests/manifest2proto-test.ts [hunk 2]
```diff
@@ -182,6 +193,187 @@ describe('manifest2proto', () => {
...
+          },
+        ]
+      },
+      {
+        specName: 'Reader1',
+        connections: [
+          {
+            name: 'first',
+            handle: 'h0',
+            type: {collection: {collectionType: {variable: {
```
Comment (piotrswigon) [inline src/tools/tests/manifest2proto-test.ts:332]: Don't we want the types on handle connection instances (i.e. real handle connections in the recipe, not the handle connections on particle specs) to be concrete types? That's what we decided is right for Plan.kt, and I think having the same decision here makes sense.

File: src/tools/tests/manifest2proto-test.ts (hunk 1 above)
Comment (alxrsngrtn) [reply src/tools/tests/manifest2proto-test.ts:255]: This is the reason that I added the null check code -- I'll fix the issue here and not patch the null.

File: src/tools/manifest2proto.ts [hunk 3]
```diff
@@ -235,24 +235,30 @@ async function recipeToProtoPayload(recipe: Recipe) {
     // all handle types are constrained type variables. We force these type variables
     // to their resolution by called maybeEnsureResolved(), so that handle types
     // are encoded with concrete types, instead of variables.
-    h.type.maybeEnsureResolved();
+    if (h.type) {
```
[the remark this answers is not in the document]
Comment (alxrsngrtn) [reply src/tools/manifest2proto.ts:238]: Found out why below, now it's no longer necessary; will rm.

File: src/tools/tests/manifest2proto-test.ts [hunk 4]
```diff
@@ -182,6 +193,187 @@ describe('manifest2proto', () => {
     assert.deepStrictEqual(connections.map(hc => hc.direction), ['READS', 'WRITES', 'READS_WRITES']);
   });
 
+  it('encodes concrete particle instance types', async () => {
```
[the remark this answers is not in the document]
Comment (alxrsngrtn) [reply src/tools/tests/manifest2proto-test.ts:196]: 👍

**Pull request merged**